pkg/gateway/connection/connections-controller.go

```diff
@@ -0,0 +1,127 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connection
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	networkingv1alpha1 "github.com/liqotech/liqo/apis/networking/v1alpha1"
+	"github.com/liqotech/liqo/pkg/gateway"
+	"github.com/liqotech/liqo/pkg/gateway/connection/conncheck"
+	"github.com/liqotech/liqo/pkg/gateway/tunnel/common"
+)
+
+// cluster-role
+// +kubebuilder:rbac:groups=networking.liqo.io,resources=connections,verbs=get;list;create;delete;update;watch
+// +kubebuilder:rbac:groups=networking.liqo.io,resources=connections/status,verbs=get;update;patch
+
+// ConnectionsReconciler updates the PublicKey resource used to establish the Wireguard connection.
+type ConnectionsReconciler struct {
+	ConnChecker    *conncheck.ConnChecker
+	Client         client.Client
+	Scheme         *runtime.Scheme
+	EventsRecorder record.EventRecorder
+	Options        *gateway.Options
+}
+
+// NewConnectionsReconciler returns a new PublicKeysReconciler.
+func NewConnectionsReconciler(ctx context.Context, cl client.Client,
+	s *runtime.Scheme, er record.EventRecorder, options *gateway.Options) (*ConnectionsReconciler, error) {
+	connchecker, err := conncheck.NewConnChecker()
+	if err != nil {
+		return nil, fmt.Errorf("unable to create the connection checker: %w", err)
+	}
+	go connchecker.RunReceiver(ctx)
+	go connchecker.RunReceiverDisconnectObserver(ctx)
+	return &ConnectionsReconciler{
+		ConnChecker:    connchecker,
+		Client:         cl,
+		Scheme:         s,
+		EventsRecorder: er,
+		Options:        options,
+	}, nil
+}
+
+// Reconcile manage PublicKey resources.
+func (r *ConnectionsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	connection := &networkingv1alpha1.Connection{}
+	if err := r.Client.Get(ctx, req.NamespacedName, connection); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("There is no connection %s", req.String())
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("unable to get the connection %q: %w", req.NamespacedName, err)
+	}
+
+	if r.ConnChecker.IsSenderRunning(r.Options.RemoteClusterID) {
+		return ctrl.Result{}, nil
+	}
+
+	remoteIP, err := common.GetRemoteInterfaceIP(r.Options.Mode)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("unable to get the remote interface IP: %w", err)
+	}
+
+	go r.ConnChecker.AddAndRunSender(
+		ctx, r.Options.RemoteClusterID, remoteIP,
+		ForgeUpdateConnectionCallback(ctx, r.Client, req),
+	)
```

Comment:
this is not guaranteed to be thread safe; the check if it is running and the set of the running one are safe